community/io/src/test/java/org/neo4j/io/pagecache/PageCacheTest.java

```diff
@@ -4862,4 +4875,434 @@ public void fileSizeMustIncreaseInPageIncriments() throws Exception
+        assertThat( filepaths, containsInAnyOrder( a.getCanonicalPath(), b.getCanonicalPath(), c.getCanonicalPath() ) );
+    }
+
+    @Test
+    public void streamFilesRecursiveMustNotListSubDirectories() throws Exception
+    {
+        configureStandardPageCache();
+        File sub1 = existingDirectory( "sub1" );
+        File sub2 = existingDirectory( "sub2" );
+        File sub2sub1 = new File( sub2, "sub1");
+        ensureDirectoryExists( sub2sub1 );
+        File sub3 = existingDirectory( "sub3" );
+        File a = existingFile( "a" );
+        File b = new File( sub1, "b" );
+        File c = new File( sub2, "c" );
+        ensureExists( b );
+        ensureExists( c );
+
+        Stream<FileHandle> stream = pageCache.streamFilesRecursive( a.getParentFile() );
+        List<String> filepaths = stream.map( FileHandle::getAbsolutePath ).collect( toList() );
+        assertThat( filepaths, containsInAnyOrder( a.getCanonicalPath(), b.getCanonicalPath(), c.getCanonicalPath() ) );
+    }
+
+    @Test
+    public void streamFilesRecursiveFilePathsMustBeCanonical() throws Exception
+    {
+        configureStandardPageCache();
+        File sub = existingDirectory( "sub" );
+        File a = new File( new File( new File( sub, ".." ), "sub" ), "a" );
+        ensureExists( a );
+
+        Stream<FileHandle> stream = pageCache.streamFilesRecursive( sub.getParentFile() );
+        List<String> filepaths = stream.map( FileHandle::getAbsolutePath ).collect( toList() );
+        assertThat( filepaths, containsInAnyOrder(
+                a.getCanonicalPath(), // file in our sub directory
+                file( "a" ).getCanonicalPath() ) ); // this file is always created by the test setup
+    }
+
+    @Test
+    public void streamFilesRecursiveMustListSingleFileGivenAsBase() throws Exception
+    {
+        configureStandardPageCache();
+        File sub = existingDirectory( "sub" );
+        File x = existingFile( "sub/x" );
+        File a = file( "a" );
+
+        Stream<FileHandle> stream = pageCache.streamFilesRecursive( a );
+        List<String> filepaths = stream.map( FileHandle::getAbsolutePath ).collect( toList() );
+        assertThat( filepaths, containsInAnyOrder( a.getCanonicalPath() ) ); // note that we don't go into 'sub'
+    }
+
+    @Test
+    public void streamFilesRecursiveListedSingleFileMustHaveCanonicalPath() throws Exception
+    {
+        configureStandardPageCache();
+        File sub = existingDirectory( "sub" );
+        File x = existingFile( "sub/x" );
+        File a = file( "a" );
+        File queryForA = new File( new File( sub, ".." ), "a" );
+
+        Stream<FileHandle> stream = pageCache.streamFilesRecursive( queryForA );
+        List<String> filepaths = stream.map( FileHandle::getAbsolutePath ).collect( toList() );
+        assertThat( filepaths, containsInAnyOrder( a.getCanonicalPath() ) ); // note that we don't go into 'sub'
+    }
+
+    @Test( expected = NoSuchFileException.class )
```

Comment:
Can we avoid using `@Test(expected)` and instead use `ExpectedException` rule or `try-catch`? Otherwise it is really hard to understand where exception is expected to be thrown from (especially true for following tests).